src/crimson/osd/exceptions.h

```diff
@@ -60,3 +60,525 @@ struct input_output_error : public error {
+                 std::forward<ErrorFuncTail>(error_func_tail)...));
+    }
+
+    template <class Func>
+    void then(Func&&) = delete;
+
+
+  private:
+    // for the sake of `plainify()` let any errorator convert errorated
+    // future into plain one.
+    template <class... AnyWrappedAllowedErrorsT>
+    friend class errorator;
+
+    base_t&& as_plain_future() && {
+      return std::move(*this);
+    }
+  };
+
+  // the visitor that forwards handling of all errors to next continuation
+  struct pass_further {
+    template <class ErrorT>
+    decltype(auto) operator()(ErrorT&&) {
+      using DecayedT = std::decay_t<ErrorT>;
+      static_assert((... || std::is_same_v<WrappedAllowedErrorsT, DecayedT>),
+                    "passing further disallowed ErrorT");
+      return DecayedT::make();
+    }
+  };
+
+  struct discard_all {
+    template <class ErrorT>
+    decltype(auto) operator()(ErrorT&&) {
+      static_assert((... || std::is_same_v<WrappedAllowedErrorsT,
+                                           std::decay_t<ErrorT>>),
+                    "discarding disallowed ErrorT");
+    }
+  };
+
+  template <class ErrorFunc>
+  static decltype(auto) all_same_way(ErrorFunc&& error_func) {
+    return [
+      error_func = std::forward<ErrorFunc>(error_func)
+    ] (auto&& e) mutable -> decltype(auto) {
+      using DecayedT = std::decay_t<decltype(e)>;
+      return DecayedT::handle(std::forward<ErrorFunc>(error_func))(e);
+    };
+  };
+
+  // get a new errorator by extending current one with new error
+  template <class... NewWrappedAllowedErrorsT>
+  using extend = errorator<WrappedAllowedErrorsT...,
+                           NewWrappedAllowedErrorsT...>;
+
+  // comparing errorators
+  template <class ErrorT>
+  struct is_carried {
+    static constexpr bool value = \
+      ((std::is_same_v<WrappedAllowedErrorsT, ErrorT>) || ...);
+  };
+  template <class ErrorT>
+  static constexpr bool is_carried_v = is_carried<ErrorT>::value;
+
+  template <class>
+  struct is_less_errorated {
+    // NOP.
+  };
+  template <class... OtherWrappedAllowedErrorsT>
+  struct is_less_errorated<errorator<OtherWrappedAllowedErrorsT...>> {
+    static constexpr bool value = \
+      ((!is_carried_v<OtherWrappedAllowedErrorsT>) || ...);
+  };
+  template <class OtherErrorator>
+  static constexpr bool is_less_errorated_v = \
+    is_less_errorated<OtherErrorator>::value;
+
+private:
+  template <class... Args>
+  static decltype(auto) plainify(seastar::future<Args...>&& fut) {
+    return std::forward<seastar::future<Args...>>(fut);
+  }
+  template <class Arg>
+  static decltype(auto) plainify(Arg&& arg) {
+    return std::forward<Arg>(arg).as_plain_future();
+  }
+
+  template <class...>
+  struct tuple2future {
+  };
+  template <class... Args>
+  struct tuple2future <std::tuple<Args...>> {
+    using type = future<Args...>;
+  };
+
+  template <class T, class = std::void_t<T>>
+  class futurize {
```

Comment:
why not `private` inherit from seastar::futurize<T>`?

Reply:
We couldn't just import those members with `using` because the `seastar::futurize` always provides the non-errorated, plain `seastar::future` while there might be a situation when:
* `ValueFunc` returns e.g. `seastar::future<>` but
* `ErrorVisitor` injects a new error.

In such case the `type` of `futurize` should be actually `errorator<new_error>::future<>`.

Reply:
This part is the reason:

```cpp
  template <class... Args>
  struct tuple2future <std::tuple<Args...>> {
    using type = future<Args...>;
  };

  template <class T, class = std::void_t<T>>
  class futurize {
    using vanilla_futurize = seastar::futurize<T>;
  public:
    using type = typename tuple2future<typename vanilla_futurize::value_type>::type;

    template <class Func, class... Args>
    static type/* can be errorated */ apply(Func&& func, std::tuple<Args...>&& args) {
```

I think `tuple2future` requires a comment emphasizing the `future<Args...>` is non-plain, errorated `future`.